Rewrite List doc (#6970)
marmelab/react-admin: docs/useList.md

---
layout: default
title: "useList"
---

# `useList`

The `useList` hook allows to create a `ListContext` based on local data. `useList` creates callbacks for sorting, paginating, filtering, and selecting records from an array.

> Thanks to it, you can display your data inside a [`<Datagrid>`](#the-datagrid-component), a [`<SimpleList>`](#the-simplelist-component) or an [`<EditableDatagrid>`](#the-editabledatagrid-component). 

Reviewer: ```suggestion
Thanks to it, you can display your data inside a [`<Datagrid>`](./Datagrid.md#datagrid), a [`<SimpleList>`](./SimpleList.md#simplelist) or an [`<EditableDatagrid>`](./EditableDatagrid.md#editabledatagrid).
```

    TextField
} from 'react-admin';

const data = [
    { id: 1, name: 'Arnold' },
    { id: 2, name: 'Sylvester' },
    { id: 3, name: 'Jean-Claude' },
];

const MyComponent = () => {
    const listContext = useList({ data });
    return (
        <ListContextProvider value={listContext}>
            <Datagrid>
                <TextField source="id" />
                <TextField source="name" />
            </Datagrid>
        </ListContextProvider>
    );
};
```

If you use it with data coming from the `dataProvider`, don't forget to pass the `isLoading` prop so that it only manipulates the data once it's available:

```jsx
import { useGetList, useList } from 'react-admin';

const MyComponent = () => {
    const { data, isLoading } = useGetList(
        'posts',
        { page: 1, perPage: 10 }
    );
    const listContext = useList({ data, isLoading });
    return (
        <ListContextProvider value={listContext}>
            <Datagrid>
                <TextField source="id" />
                <TextField source="name" />
            </Datagrid>
        </ListContextProvider>
    );
};
```

The `useGetList` parameter accepts the following options:

* [`filter`](#filter)

> * [`isFetching`](#isFetching)

Reviewer: ```suggestion
* [`isFetching`](#isfetching)
```

> * [`isLoading`](#isLoading)

Reviewer: ```suggestion
* [`isLoading`](#isloading)
```

* [`page`](#page)

> * [`perPage`](#perPage)

Reviewer: ```suggestion
* [`perPage`](#perpage)
```

* [`sort`](#sort)

## `filter`

The initial filter to apply to the data.

```jsx
const { data, total } = useList({
    data: [
        { id: 1, name: 'Arnold' },
        { id: 2, name: 'Sylvester' },
        { id: 3, name: 'Jean-Claude' },
    ],
    filter: { name: 'Arnold' },
});
// data will be [{ id: 1, name: 'Arnold' }] and total will be 1
```

## `isFetching`

This value ends up in the return value. It is used by list iterators (like `<Datagrid>`) to know when to display a loading indicator.

```jsx
import { useGetList, useList } from 'react-admin';

const MyComponent = () => {
    const { data, isFetching } = useGetList(
        'posts',
        { page: 1, perPage: 10 }
    );
    const listContext = useList({ data, isFetching });
    return (
        <ListContextProvider value={listContext}>
            <Datagrid>
                <TextField source="id" />
                <TextField source="name" />
            </Datagrid>
        </ListContextProvider>
    );
};
```

## `isLoading`

This value ends up in the return value. It is used by list iterators (like `<Datagrid>`) to know when to display a loading indicator.

```jsx

> import { useGetList, useList } from 'react-admin';

Reviewer: ```suggestion
import { useGetList, useList, Datagrid, ListContextProvider, TextField } from 'react-admin';
```
